web/docs/develop-vast/write-a-plugin.md

```diff
@@ -0,0 +1,256 @@
+builds.
+
+VAST ships with an example plugin that showcases how a typical scaffold looks
+like. Have a look at the the [example
+plugins](https://github.com/tenzir/vast/tree/master/examples/plugins) directory,
+and an [example `CMakeLists.txt` file for
+plugins](https://github.com/tenzir/vast/blob/master/examples/plugins/analyzer/CMakeLists.txt).
+
+We recommend calling the provided `VASTRegisterPlugin` CMake in your plugin's
+`CMakeLists.txt` file instead of handrolling your CMake build scaffolding
+code. This ensures that your plugin always uses the recommended defaults.
+Non-static installations of VAST contain the `VASTRegisterPlugin.cmake` modules.
+
+If present in the plugin source directory, the build scaffolding automatically
+installs `<plugin>.yaml.example` files, commenting out every line so the file
+has no effect. This serves as documentation for operators that can modify the
+installed file in-place.
+
+Similarly, the build scaffolding installs files in the `schema` subdirectory
+automatically such that loading the plugin loads additionally required schema
+files automatically.
+
+## Choose a plugin type
+
+VAST offers a variety of customization points, each of which defines its own
+API by inheriting from the plugin base class `vast::plugin`. When writing a new
+plugin, you can choose a subset of available types by inheriting from
+the respective plugin classes.
+
+:::caution Dreaded Diamond
+To avoid common issues with multiple inheritance, all intermediate plugin
+classes that inherit from `vast::plugin` use *virtual inheritance* to avoid
+issues with the [dreaded
+diamond](https://isocpp.org/wiki/faq/multiple-inheritance#mi-diamond).
+:::
+
+The below section describes the [various plugin types](#plugin-types) in more
+detail. Please also consult the [example-analyzer
+plugin](https://github.com/tenzir/vast/tree/master/examples/plugins/analyzer)
+for a complete end-to-end code example.
+
+## Implement the plugin interface
+
+After having the necessary CMake in place, you can now derive from one or more
+plugin base classes to define your own plugin. Based on the chosen plugin
+types, you must override one or more virtual functions with an implementation
+of your own.
+
+The basic anatomy of a plugin class looks as follows:
+
+```cpp
+class example_plugin final : public virtual analyzer_plugin,
+                             public virtual command_plugin {
+public:
+  /// Loading logic.
+  example_plugin();
+
+  /// Teardown logic.
+  ~example_plugin() override;
+
+  /// Initializes a plugin with its respective entries from the YAML config
+  /// file, i.e., `plugin.<NAME>`.
+  /// @param config The relevant subsection of the configuration.
+  caf::error initialize(data config) override;
+
+  /// Returns the unique name of the plugin.
+  const char* name() const override;
+
+  // TODO: override pure virtual functions from the base classes.
+  // ...
+
+private:
+  record config_;
+};
+```
```

Comment:
This should just link to the example analyzer plugin instead.

Reply:
I actually like this, because it shows the compactness of the plugin class. Strong opinions on the linking?

Reply:
We resolved this in a discussion.